rabbitmq/rabbitmq-stream-dotnet-client
# Improve Reconnection

* Part of https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/issues/336

* Improve the disconnection message. It is possible to understand if it is a normal disconnection or an unexpected.
* Remove the Active Items from the connection pool and use the Publishers and Consumers client list directly to check the pool size
* Refactor the Factory Classes. Remove code duplication in case of metadata update and connection closed. See ReliableBase.OnEntityClosed
* Handle streamNotAvailable error. In this case the client will try to reconnect the entity
* Fix the events attach to the RawConsumer and RawProducer. The events are attached only if the ResponseCode is OK
* Detach the events once the entity receives the disconnection or metadata update. In that case, the entity is closed
* Introduce ReliableEntityStatus like a state machine to understand the status of Producer and Consumer classes
* Add ResourceAvailableReconnectStrategy interface to Handle the retry in case testing in a stream exists. See ReliableBase CheckIfStreamIsAvailable
* Change the  MetadataHandler to Func<MetaDataUpdate, Task> to be like the other methods
* Producer and Consumer classes fail fast during the first initialization. The user is aware of what is happening. The reconnect part occurs only after the first boot.



How to test
----
- Setup a cluster. I use:
```
git clone git@github.com:rabbitmq/rabbitmq-stream-go-client.git .
make rabbitmq-ha-proxy 
```

```
/etc/hosts
127.0.0.1 node0
127.0.0.1 node1
127.0.0.1 node2
```

- execute a script with multiple consumers and producers:
I use https://gist.github.com/Gsantomaggio/f6a091076e498d88ecfe827c2b165dad


- restart the docker nodes and see if the client will survive to the restart
